### PR TITLE
feat: support Angular 12

### PR DESCRIPTION
**Pull request by arturovt**

### CODE REVIEW

**Comment by arturovt** [comment]
@joeldenning cannot make it working since it throws:

> Error: You forgot to add 'mini-css-extract-plugin' plugin (i.e. `{ plugins: [new MiniCssExtractPlugin()] }`), please read https://github.com/webpack-contrib/mini-css-extract-plugin#getting-started

https://stackoverflow.com/questions/65961356/how-to-get-mini-css-extract-plugin-to-work-with-webpack-5

Do you have any ideas?

**Comment by joeldenning** [comment]
We're currently removing mini-css-extract-plugin from the webpack config in favor of using style-loader, since it's hard to make sure the css is loaded from the correct url at the correct time with mini-css-extract-plugin. I created https://github.com/single-spa/single-spa-css to help with this, but haven't fully tested or incorporated it into react/vue/angular single-spa CLIs yet.

Does angular 12 just refuse to compile without that plugin? If so, first attempt would be to work around that. Second attempt would be to use single-spa-css in conjunction with single-spa-angular to ensure proper loading of css files.

**Comment by joeldenning** [comment]
Oh I just realized a more likely explanation - it's probably not Angular, but the mini-css-extract-plugin, that is throwing that error. The plugin requires a webpack loader and a webpack plugin - perhaps single-spa-angular is removing the plugin without removing the loader. See https://www.npmjs.com/package/mini-css-extract-plugin for docs. Sharing the relevant code in single-spa-angular below - my guess is that angular 12's webpack config is different such that we no longer enter the `if` statement on line 104

https://github.com/single-spa/single-spa-angular/blob/9c6b8fcf23fc5f167180608bba27309b9686f4d0/libs/single-spa-angular/webpack/index.ts#L94-L108

**Comment by larsf96** [comment]
I've just tested the current single-spa-angular version with Angular 12.
One error that showed up was:
```
Error: node_modules/single-spa-angular/src/extra-providers.d.ts:12:5 - error TS2416: Property 'onPopState' in type 'SingleSpaPlatformLocation' is not assignable to the same property in base type 'ɵBrowserPlatformLocation'.
  Type '(fn: OnPopStateListener) => void' is not assignable to type '(fn: LocationChangeListener) => VoidFunction'.
    Type 'void' is not assignable to type 'VoidFunction'.
```
After changing the type of "onPopState" from void to VoidFunction, the error is gone

Another point is that I had to remove the aot=true option from the angular.json file, otherwise some errors showed up:
```
Error: ./src/main.single-spa.ts
Module build failed (from ./node_modules/@ngtools/webpack/src/ivy/index.js):
TypeError: angularCompiler.getNextProgram is not a function
```

I'm not sure if this has something to do with single-spa-angular at all, looks more like an angular issue I guess.

File: libs/single-spa-angular/internals/src/dom.ts [hunk 1]
```diff
@@ -1,19 +1,5 @@
 import { DomElementGetter, BaseSingleSpaAngularOptions } from './types';
 
-export function removeApplicationFromDOMIfIvyEnabled<T extends BaseSingleSpaAngularOptions>(
```
Comment (joeldenning) [inline libs/single-spa-angular/internals/src/dom.ts:3]: This makes this a breaking change, which is okay. We can publish a breaking version for Angular 12

File: libs/single-spa-angular/src/extra-providers.ts [hunk 2]
```diff
@@ -16,50 +15,7 @@ export class SingleSpaPlatformLocation extends ɵBrowserPlatformLocation {
...
-  // When shop application gets destroyed we want to remove only its `popstate` listener.
-  private zoneToOnPopStateListenersMap = new Map<any, OnPopStateListener[]>();
-
-  // This is used only to make `Zone.wrap` happy, since it requires 2 arguments
-  // and the second argument is a unique string which `zone.js` uses for debugging purposes.
-  // We might want to use the application name, but we're not able to get it when `onPopState`
-  // method is called during module bootstrapping.
-  private source = 0;
-
-  destroyApplication(zoneIdentifier: string): void {
```
Comment (joeldenning) [inline libs/single-spa-angular/src/extra-providers.ts:38]: Did Angular 12 fix the underlying bug that required this? If so, link to pull request?

File: libs/single-spa-angular/src/single-spa-angular.ts [hunk 3]
```diff
@@ -151,52 +141,15 @@ async function mount(options: SingleSpaAngularOptions, props: any): Promise<NgMo
...
-    // Workaround for https://github.com/angular/angular/issues/19079
-    const router = options.bootstrappedModule!.injector.get(options.Router);
-    router.dispose();
-  }
-
-  if (options.routingEventListener) {
-    window.removeEventListener('single-spa:routing-event', options.routingEventListener);
-  }
-
-  if (options.AnimationEngine) {
```
Comment (joeldenning) [inline libs/single-spa-angular/src/single-spa-angular.ts:166]: Does AnimationEngine now clean itself up properly during ngDestroy?

File: migrations.json [hunk 4]
```diff
@@ -0,0 +1,108 @@
+{
+  "migrations": [
```
Comment (joeldenning) [inline migrations.json:2]: What is this file for? Is it related to schematics, or to nx?

**Review by joeldenning — COMMENTED** [review]
A few questions

File: libs/single-spa-angular/src/extra-providers.ts (hunk 2 above)
Comment (arturovt) [reply libs/single-spa-angular/src/extra-providers.ts:38]: https://github.com/angular/angular/pull/40867

File: libs/single-spa-angular/src/single-spa-angular.ts (hunk 3 above)
Comment (arturovt) [reply libs/single-spa-angular/src/single-spa-angular.ts:166]: https://github.com/angular/angular/pull/41059

File: migrations.json (hunk 4 above)
Comment (arturovt) [reply migrations.json:2]: It's auto generated by Nx.

**Review by joeldenning — APPROVED** [review]
looks good - I'll merge and publish a new version later today or tomorrow. If I forget, feel free to remind me here to do it.

**Comment by patelvimal** [comment]
> looks good - I'll merge and publish a new version later today or tomorrow. If I forget, feel free to remind me here to do it.

Reminder, please publish the new version, waiting for new version.

**Comment by joeldenning** [comment]
Am having trouble with the publish - `yarn build` does not create the directories and files I would expect. No package.json nor the main single-spa-angular.js bundles. Instead, it's creating the files below. I'll investigate further. @arturovt do you know what would cause this?

```
lib
lib/README.md
lib/lib
lib/lib/webpack
lib/lib/webpack/webpack-5
lib/lib/webpack/webpack-5/remove-mini-css-extract.d.ts
lib/lib/webpack/webpack-5/remove-mini-css-extract.js.map
lib/lib/webpack/webpack-5/remove-mini-css-extract.js
lib/lib/webpack/index.js
lib/lib/webpack/index.js.map
lib/lib/webpack/index.d.ts
lib/schematics
lib/schematics/schematics.json
lib/schematics/ng-add
lib/schematics/ng-add/tests
lib/schematics/ng-add/tests/utils.d.ts
lib/schematics/ng-add/tests/utils.js
lib/schematics/ng-add/tests/utils.js.map
lib/schematics/ng-add/index.js
lib/schematics/ng-add/rules
lib/schematics/ng-add/rules/update-configuration.d.ts
lib/schematics/ng-add/rules/add-dependencies.js.map
lib/schematics/ng-add/rules/update-configuration.js.map
lib/schematics/ng-add/rules/show-warning-if-routing-is-enabled.d.ts
lib/schematics/ng-add/rules/add-npm-scripts.js
lib/schematics/ng-add/rules/add-npm-scripts.d.ts
lib/schematics/ng-add/rules/add-dependencies.js
lib/schematics/ng-add/rules/show-warning-if-routing-is-enabled.js.map
lib/schematics/ng-add/rules/create-main-entry.js.map
lib/schematics/ng-add/rules/utils.d.ts
lib/schematics/ng-add/rules/update-configuration.js
lib/schematics/ng-add/rules/add-npm-scripts.js.map
lib/schematics/ng-add/rules/utils.js
lib/schematics/ng-add/rules/utils.js.map
lib/schematics/ng-add/rules/create-main-entry.js
lib/schematics/ng-add/rules/create-main-entry.d.ts
lib/schematics/ng-add/rules/show-warning-if-routing-is-enabled.js
lib/schematics/ng-add/rules/add-dependencies.d.ts
lib/schematics/ng-add/index.js.map
lib/schematics/ng-add/schema.json
lib/schematics/ng-add/schema.d.ts
lib/schematics/ng-add/index.d.ts
lib/schematics/ng-add/_files
lib/schematics/ng-add/_files/extra-webpack.config.js.template
lib/schematics/ng-add/_files/src
lib/schematics/ng-add/_files/src/app
lib/schematics/ng-add/_files/src/app/empty-route
lib/schematics/ng-add/_files/src/app/empty-route/empty-route.component.ts.template
lib/schematics/ng-add/_files/src/main.single-spa.ts.template
lib/schematics/ng-add/_files/src/single-spa
lib/schematics/ng-add/_files/src/single-spa/asset-url.ts.template
lib/schematics/ng-add/_files/src/single-spa/single-spa-props.ts.template
```

**Comment by joeldenning** [comment]
Seems to be a problem with the nx build inside of `build:single-spa-angular`. Here's the terminal output, but no files are created:

```
yarn build:single-spa-angular
yarn run v1.22.10
warning ../../package.json: No license field
$ nx build single-spa-angular && cpx README.md lib

> nx run single-spa-angular:build [existing outputs match the cache, left as is]
Building Angular Package

------------------------------------------------------------------------------
Building entry point 'single-spa-angular/internals'
------------------------------------------------------------------------------

------------------------------------------------------------------------------
Building entry point 'single-spa-angular/parcel'
------------------------------------------------------------------------------

------------------------------------------------------------------------------
Building entry point 'single-spa-angular'
------------------------------------------------------------------------------

------------------------------------------------------------------------------
Building entry point 'single-spa-angular/elements'
------------------------------------------------------------------------------

------------------------------------------------------------------------------
Built Angular Package
 - from: /Users/joeldenning/code/single-spa-angular/libs/single-spa-angular
 - to:   /Users/joeldenning/code/single-spa-angular/lib
------------------------------------------------------------------------------

Build at: 2021-06-01T21:34:26.804Z - Time: 8733ms


———————————————————————————————————————————————

>  NX   SUCCESS  Running target "build" succeeded

  Nx read the output from cache instead of running the command for 1 out of 1 tasks.


>  NX   CLOUD  See run details at https://nx.app/runs/uiU0OJewVRc

✨  Done in 1.86s.
```

```sh
$ find lib
lib
lib/README.md
```

**Comment by arturovt** [comment]
@joeldenning I've pushed to master branch, could you pull and try again

**Comment by joeldenning** [comment]
That worked! Thanks. Publishing now. I've created the [4.x branch](https://github.com/single-spa/single-spa-angular/tree/4.x) for long term maintenance of the 4.x releases

**Comment by joeldenning** [comment]
Released in https://github.com/single-spa/single-spa-angular/releases/tag/v5.0.0.

**Comment by joeldenning** [comment]
Documented in https://github.com/single-spa/single-spa.js.org/pull/456

**Comment by larsf96** [comment]
I've just tried the new version with Angular 12, but I'm getting an error when using "ng serve"

```
 - configuration.output has an unknown property 'chunkLoadingGlobal'. These properties are valid:
   object { auxiliaryComment?, chunkCallbackName?, chunkFilename?, chunkLoadTimeout?, crossOriginLoading?, devtoolFallbackModuleFilenameTemplate?, devtoolLineToLine?, devtoolModuleFilenameTemplate?, devtoolNamespace?, filename?, futureEmitAssets?, globalObject?, hashDigest?, hashDigestLength?, hashFunction?, hashSalt?, hotUpdateChunkFilename?, hotUpdateFunction?, hotUpdateMainFilename?, jsonpFunction?, jsonpScriptType?, library?, libraryExport?, libraryTarget?, path?, pathinfo?, publicPath?, sourceMapFilename?, sourcePrefix?, strictModuleExceptionHandling?, umdNamedDefine?, webassemblyModuleFilename? }
   -> Options affecting the output of the compilation. `output` options tell webpack how to write the compiled files to disk.
 
```

**Pull request merged**